### PR TITLE
[Doc] Improve ListActions override

### PR DESCRIPTION
**Pull request by WiXSL**
Implements #6212

### CODE REVIEW

**Comment by djhi** [comment]
:+1:  Maybe we should do the same for the other action components as well.

**Review by fzaninotto — COMMENTED** [review]
Thanks for your patch. However, I'm not comfortable with this change, and for 2 reasons:

- it creates 2 ways to do one simple thing. This is confusing, which way should a developer use and why? 
- your API "hides" that some other buttons are included. I think that will lead to WTF moments for developers. 

Overall, I don't think we should implement that.

**Comment by WiXSL** [comment]
Ok, how about move the logic of the actions to hooks and you always override the actions prop like now, but having control of default actions by calling this hooks (or not).

Example: 

```jsx
import Button from '@material-ui/core/Button';
import { TopToolbar, Edit, useEditDefaultActions } from 'react-admin';

const PostEditActions = (props) => {
     const DefaultActions = useEditDefaultActions(props);

    return <TopToolbar>
        <DefaultActions/> 
        {/* Add your custom actions */}
        <Button 
            color="primary"
            onClick={() => { alert('Your custom action'); }}
        >
            Custom Action
        </Button>
    </TopToolbar>;
};

export const PostEdit = (props) => (
    <Edit actions={<PostEditActions />} {...props}>
        ...
    </Edit>
);
```

**Comment by WiXSL** [comment]
@fzaninotto,
What do you think of this idea?

**Comment by djhi** [comment]
Why a hook here? We could just have a component rendering a fragment with the default actions right?

**Comment by WiXSL** [comment]
You are right.

**Comment by fzaninotto** [comment]
When I think about it, only the `<ListActions>` are hard to override. Maybe this is where we should do the change. 

First, the documentation could be simplified since we now use the various context:

```diff
const ListActions = (props) => {
    const {
        className,
-       exporter,
-       filters,
-       maxResults,
        ...rest
    } = props;
    const {
-       currentSort,
-       resource,
-       displayedFilters,
-       filterValues,
-       hasCreate,
-       basePath,
-       selectedIds,
-       showFilter,
        total,
    } = useListContext();
    return (
        <TopToolbar className={className} {...sanitizeListRestProps(rest)}>
-           {filters && cloneElement(filters, {
-               resource,
-               showFilter,
-               displayedFilters,
-               filterValues,
-               context: 'button',
-           })}
+           {cloneElement(filters, { context: 'button' })}
-           <CreateButton basePath={basePath} />
+           <CreateButton />
-           <ExportButton
-               disabled={total === 0}
-               resource={resource}
-               sort={currentSort}
-               filterValues={filterValues}
-               maxResults={maxResults}
-           />
+           <ExportButton disabled={total === 0} maxResults={maxResults} />
            {/* Add your custom actions */}
            <Button
                onClick={() => { alert('Your custom action'); }}
                label="Show calendar"
            >
                <IconEvent />
            </Button>
        </TopToolbar>
    );
};
```

For a simpler use case, the doc is even simpler:

```jsx
const ListActions = () => (
    <TopToolbar >
        {cloneElement(filters, { context: 'button' })}
        <CreateButton />
        <ExportButton />
        {/* Add your custom actions */}
        <Button
            onClick={() => { alert('Your custom action'); }}
            label="Show calendar"
        >
            <IconEvent />
        </Button>
    </TopToolbar>
);
```

I think this is simple enough that we don't have to introduce a new component for that.

**Comment by WiXSL** [comment]
So, I should revert everything and update the only documentation?

File: docs/List.md [hunk 1]
```diff
@@ -90,7 +90,37 @@ The title can be either a string or an element of your own.
 
 ![Actions Toolbar](./img/actions-toolbar.png)
 
-You can replace the list of default actions by your own element using the `actions` prop:
+You can add extra actions to a `<List>` by using the `actions` prop:
```
Comment (djhi) [inline docs/List.md:93]: The original sentence was fine imo

File: docs/List.md [hunk 2]
```diff
@@ -90,7 +90,37 @@ The title can be either a string or an element of your own.
...
+);
+
+export const PostList = (props) => (
+        <List {...props} actions={<ListActions/>}>
+          ...
+        </List>
+);
+```
+
+If you want to have control over default actions as well, you can replace the entire `actions` component by your own:
```
Comment (djhi) [inline docs/List.md:123]:
```suggestion
This allows you to further control how the default actions behaves. For example, you could disable the `<ExportButton>` when the list is empty:
```

**Review by djhi — CHANGES_REQUESTED** [review]
Thanks!

File: docs/List.md [hunk 3]
```diff
@@ -92,6 +92,36 @@ The title can be either a string or an element of your own.
...
+);
+
+export const PostList = (props) => (
+        <List {...props} actions={<ListActions/>}>
+          ...
+        </List>
+);
+```
+
+This allows you to further control how the default actions behaves. For example, you could disable the `<ExportButton>` when the list is empty:
```
Comment (djhi) [inline docs/List.md:123]:
```suggestion
This allows you to further control how the default actions behave. For example, you could disable the `<ExportButton>` when the list is empty:
```

Achievement unlocked: tricked El Corrector :laughing:

**Review by djhi — CHANGES_REQUESTED** [review]

**Review by djhi — APPROVED** [review]

File: docs/List.md [hunk 4]
```diff
@@ -104,45 +134,25 @@ import {
     ExportButton,
     Button,
     sanitizeListRestProps,
+    useListContext
```
Comment (fzaninotto) [inline docs/List.md:137]: useListContext is already included line 131

File: packages/ra-ui-materialui/src/list/ListActions.tsx [hunk 5]
```diff
@@ -15,6 +15,47 @@ import { ToolbarProps } from '@material-ui/core';
...
+ *          basePath,
+ *          selectedIds,
+ *          showFilter,
+ *          total,
+ *     } = useListContext();
+ *
+ *     const PostListActions = ({ basePath }) => (
+ *         <TopToolbar>
+ *             <CreateButton basePath={basePath} />
+ *             <ExportButton
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/ListActions.tsx:42]: why  not use the simpler version?

```
{cloneElement(props.filters, { context: 'button' })}  
<CreateButton/>
<ExportButton/>
```

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**